src/compress.c

```diff
@@ -0,0 +1,107 @@
+// Copyright (C) 2002-2006 Andrew Tridgell
+// Copyright (C) 2009-2018 Joel Rosdahl
+//
+// This program is free software; you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program; if not, write to the Free Software Foundation, Inc., 51
+// Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#include "ccache.h"
+
+#include "manifest.h"
+#include "result.h"
+
+static unsigned num_files;
+static unsigned comp_files;
+
+static uint64_t cache_size;
+static uint64_t real_size;
+
+// This measures the size of files in the cache.
+static void
+measure_fn(const char *fname, struct stat *st)
+{
+	if (!S_ISREG(st->st_mode)) {
+		return;
+	}
+
+	char *p = basename(fname);
+	if (str_eq(p, "stats")) {
+		goto out;
+	}
+
+	if (str_startswith(p, ".nfs")) {
+		// Ignore temporary NFS files that may be left for open but deleted files.
+		goto out;
+	}
+
+	if (strstr(p, "CACHEDIR.TAG")) {
+		goto out;
+	}
+
+	size_t uncompressed_size;
+	bool is_compressed;
+	if (str_endswith(p, ".manifest")) {
+		uncompressed_size = manifest_size(fname, &is_compressed);
```

Comment:
I now realize that I should have mentioned that there is a better way of doing compression statistics and recompression now. Sorry.

The file formats and "common header" functions have already been prepared to make this use case (compression statistics and recompression) efficient, see 5905237ebb917bcb1bb773431f8e8703899194cb and 318f980eb9e7424ccd166ee3de795beb0d355984 (also pushed 7839217088c7 to master just now). Since there is a content size field in the header, there's no need to dry-run decompression to calculate the content size.

So for compression inspection something like this can be done:

```c
FILE *f = fopen(path, "r");
struct common_header header;
common_header_initialize_for_reading(
	&header,
	f,
	magic, // MANIFEST_MAGIC or RESULT_MAGIC,
	version, // MANIFEST_VERSION or RESULT_VERSION,
	NULL,
	NULL,
	NULL);
// Use header.content_size for uncompressed size and stat(path, ...) for actual size.
fclose(f);
```

For recompression:

1. Use `common_header_initialize_for_reading` to read the header but with non-NULL `decompressor` and `decompressor_state` arguments.
2. If `header.compression_type` and `header.compression_level` are what we want, we're done. Otherwise:
3. Open a new file and use `common_header_initialize_for_writing`.
4. Feed data from `decompressor.read` to `compressor.write`. No need to know anything about the actual body format.

Reply:
Done, recompression is left to do (not in this PR)